### ISSUE TITLE
 primitive-0.6.2.0 doesn't compile with ghc 8.4.1 alpha

### ISSUE THREAD

**Issue by GeorgeCo** [haskell/primitive#71]
In case anybody wants to get started with this
 primitive-0.6.2.0 doesn't compile with ghc 8.4.1 alpha

```
 cabal install -j5 primitive --allow-newer
Resolving dependencies...
Configuring primitive-0.6.2.0...
Building primitive-0.6.2.0...
Failed to install primitive-0.6.2.0
Build log ( /Users/gcolpitts/.cabal/logs/ghc-8.4.0.20171214/primitive-0.6.2.0-GyZXr8FnXgfJdktXPA3hr1.log ):
cabal: Entering directory '/var/folders/9b/rh4y2gy92hgdb6ktv4df1jv00000gn/T/cabal-tmp-36646/primitive-0.6.2.0'
Configuring primitive-0.6.2.0...
Preprocessing library for primitive-0.6.2.0..
Building library for primitive-0.6.2.0..
[ 1 of 12] Compiling Control.Monad.Primitive ( Control/Monad/Primitive.hs, dist/build/Control/Monad/Primitive.o )

Control/Monad/Primitive.hs:45:1: warning: [-Wdeprecations]
    Module ‘Control.Monad.Trans.List’ is deprecated:
      This transformer is invalid on most monads
   |
45 | import Control.Monad.Trans.List     ( ListT    )
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Control/Monad/Primitive.hs:47:1: warning: [-Wdeprecations]
    Module ‘Control.Monad.Trans.Error’ is deprecated:
      Use Control.Monad.Trans.Except instead
   |
47 | import Control.Monad.Trans.Error    ( ErrorT, Error)
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Control/Monad/Primitive.hs:101:36: warning: [-Wdeprecations]
    In the use of type constructor or class ‘ListT’
    (imported from Control.Monad.Trans.List):
    Deprecated: "This transformer is invalid on most monads"
    |
101 | instance PrimMonad m => PrimMonad (ListT m) where
    |                                    ^^^^^

Control/Monad/Primitive.hs:102:19: warning: [-Wdeprecations]
    In the use of type constructor or class ‘ListT’
    (imported from Control.Monad.Trans.List):
    Deprecated: "This transformer is invalid on most monads"
    |
102 |   type PrimState (ListT m) = PrimState m
    |                   ^^^^^

Control/Monad/Primitive.hs:109:11: warning: [-Wdeprecations]
    In the use of type constructor or class ‘Error’
    (imported from Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"
    |
109 | instance (Error e, PrimMonad m) => PrimMonad (ErrorT e m) where
    |           ^^^^^

Control/Monad/Primitive.hs:109:47: warning: [-Wdeprecations]
    In the use of type constructor or class ‘ErrorT’
    (imported from Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"
    |
109 | instance (Error e, PrimMonad m) => PrimMonad (ErrorT e m) where
    |                                               ^^^^^^

Control/Monad/Primitive.hs:110:19: warning: [-Wdeprecations]
    In the use of type constructor or class ‘ErrorT’
    (imported from Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"
    |
110 |   type PrimState (ErrorT e m) = PrimState m
    |                   ^^^^^^
[ 2 of 12] Compiling Data.Primitive.Internal.Compat ( Data/Primitive/Internal/Compat.hs, dist/build/Data/Primitive/Internal/Compat.o )
[ 3 of 12] Compiling Data.Primitive.Array ( Data/Primitive/Array.hs, dist/build/Data/Primitive/Array.o )

Data/Primitive/Array.hs:531:10: error:
    • No instance for (Semigroup (Array a))
        arising from the superclasses of an instance declaration
    • In the instance declaration for ‘Monoid (Array a)’
    |
531 | instance Monoid (Array a) where
    |          ^^^^^^^^^^^^^^^^
cabal: Leaving directory '/var/folders/9b/rh4y2gy92hgdb6ktv4df1jv00000gn/T/cabal-tmp-36646/primitive-0.6.2.0'
cabal: Error: some packages failed to install:
primitive-0.6.2.0-GyZXr8FnXgfJdktXPA3hr1 failed during the building phase. The
exception was:
ExitFailure 1
```

**Comment by cartazio** [comment]
Use hackage head.  It has everything building for head

**Comment by ndmitchell** [comment]
@cartazio - why not release this package to Hackage now, so people can just use hackage? I personally don't know what it takes to use Hackage head, and have little enthusiasm for plumbing it through all my CI scripts...

**Comment by svenpanne** [comment]
I would very much appreciate a Hackage release working with GHC 8.4.1 alpha, too. `primitive` has a lot of reverse dependencies (224 direct ones), even more if you consider them transitively. As a consequence, quite a few popular packages like `QuickCheck` are broken right now with 8.4.1.

I don't think it's a sensible solution to tell hundreds of dependent projects "Hey, just fiddle around with your CI scripts...". ;-) `primitive` is very near to the bottom of the whole Hackage dependecy tree, so it would be a pity if it delayed tons of other projects. IMHO the whole point of the GHC alpha releases is to fix such breakage early, so we have a healthy ecosystem when the final release comes out.

**Comment by cartazio** [comment]
I’ll check with the rest of the maintainers and we’ll make sure to gift yah
with something nice for Xmas

@hvr and Ryan: any thoughts ?

**Comment by cartazio** [comment]
https://hackage.haskell.org/package/primitive/maintainers/

Herbert and Dan are the two active uploaders at present

**Comment by hvr** [comment]
GHC 8.4.1 is currently still a moving target and it's too early to release to Hackage (and I personally advise strongly against; Hackage releases are intended to be durable). Fwiw, I devised http://HEAD.hackage.haskell.org/ so we wouldn't need to make releases for an unreleased GHC, even less so for an **alpha** release for which GHC HQ gives no API stability guarantees whatsoever. I for one won't make any releases specifically for GHC 8.4.1 support before the first release candidate of GHC 8.4.1 hasn't been cut (and thus its API is frozen).

**Comment by cartazio** [comment]
Agreed and better said than I :)

**Comment by svenpanne** [comment]
I'm less than satisfied with this attitude: Do you really think it is a good idea to tell hundreds of project to change their CI? What should people using stack do? Is it really *that* hard for a relatively small project at the bottom of the food chain to make a new release? It's totally clear that GHC's APIs are not frozen yet, but even if they change before the final release: Just make a small fix and release again.

Given this resistance, I really have a hard time seeing the value of GHC alpha releases...

**Comment by cartazio** [comment]
when did these hundreds of projects add alpha release ghc to their CI? either way its an alpha release and its high holidays in several continents, be patient and enjoy time with friends and fam.  anyone testing software out when its hot off of alpha tags can be a tad patient :) 

merry xmas

**Comment by ndmitchell** [comment]
Are you saying "this package will be released after Christmas" or "this package will be released once GHC is in beta"? Is GHC HQ saying ["put this alpha through its paces"](https://mail.haskell.org/pipermail/ghc-devs/2017-December/015235.html) or ["it's too early"](https://github.com/haskell/primitive/issues/71#issuecomment-353664000)?

CC @mchakravarty, as I feel this impacts the GHC release cycle speed quite a bit (still working out the kinks in the new process, so no great shock at this stage!)

Happy Christmas!

**Comment by phadej** [comment]
Doesn't `stack-1.6.1` support custom snapshots. Could stack users collaborate and make a custom snapshot with git dependencies to "supposed to work with GHC-8.4.1-alpha" commits?

**Comment by ndmitchell** [comment]
@phadej: yes, they could, it’s a great time idea - but probably orthogonal (I certainly don’t use stack for alpha testing)

**Comment by cartazio** [comment]
CAbal supports multiple package dbs. What’s wrong with hackage head?

**Comment by treeowl** [comment]
@hvr, I understand your general point about API instability, but in this particular case it's absurd. The API change that is holding this package down (implementing the `Semigroup => Monoid` proposal) is one that was formally proposed on the library list and formally accepted and implemented. It is *not* going to be reversed before the `base-4.11` release. Moreover, even if it *were* reversed, there is *no downside* to supplying `Semigroup` instances as required. It is true that the way the `Monoid` instances were changed in 7e6e7b4667020b61986c60e0c2d642e700e7966d is incompatible with (hypothetically) reverting that change. Making it compatible would be a matter of adding `mappend = (<>)` where required. This issue is annoying everyone; please fix it.

**Comment by vincenthz** [comment]
@phadej not that the issue has anything to do with stack but ... considering the amount of work for creating and maintaining (tracking all git repo update) a stack snapshot for something that is ephemeral, I think the likely outcome is that it won't happen, lowering the amount of testing GHC alpha gets and the feedback GHC Devs gets back. Even if that stack snapshot happens, who will sweat to add this to their travis which will have to be undone once 8.4 is released ? I think if you don't use the same mechanism of delivery for everything then you're making the testing bar way too high already.

Clearly if there's no release of `primitive` (and other basic packages), I don't think anyone should bother with alpha or beta releases (haskellers or GHC devs) compared to say, testing GHC git repo directly.

**Comment by cartazio** [comment]
I’ll see about getting primitive in shape in a few days.  Right now this thread isn’t helping in a positive way.

**Comment by cartazio** [comment]
I’ve locked the thread to prevent more drive by non constructive remarks.  

Positive reinforcement and collaboration are way more effective way to motivate folks, and honestly framing your heated remarks in terms of personal feelings or motivations would be more constructive than how some posters have approached it. 

Hypothetical positive remark “I would really like to test ghc 8.4 alpha with vanilla hackage and a release of primitive , because for structural reasons I can’t edit my cabal config in my homedir  to point to two package sources even though that’s supported. The reasons why I can’t include xyz.  

Additionally I want to make sure my software works on day zero of the final ghc 8.4 release and better sooner than later :)“

I might be totally wrong on that last point but I do believe you can have your cabal config in your home dir point 

Either way, valid perspectives have been articulated, but drive by commenting isn’t actionable, and if not done in the right way can be demotivating. 

Vincent: your comment did not add anything constructive or new. Shame on you.  You just did a more confrontational exaggerated version of what was already reasonably articulated by Neil and David. 

Happy New Years all, I’m off to visit my younger sister for a week of rest.  Use positivity and positive reinforcement to motivate folks. And in the future please  explain why tools like hackage head aren’t effective options for you to test libraries on an alpha quality compiler release please.  Explaining WHY in terms of actions or work flows or implicit needs we don’t know about is helpful.

**Comment by vincenthz** [comment]
nice tone policing + psych 101 course + insults from cartazio.

Now can you let an actual *real* maintainer (close/wontfix/ignore/resolve) this issue ? I'ld like to know if we can expect to be able to test with non-final GHC 8.4; I personally have tried many time in the past to add non final ghc to my package builds and be stuck with unreleased primitive (or other packages), just like people commenting and watching this issue.

**Comment by ndmitchell** [comment]
@vincenthz I'm not sure your comments are driving this conversation in the right direction. You might reasonably argue it's not in the place it should be, and I appreciate the temptation to vent, but let's not make that this issue.

@cartazio This bug refers to an issue that is still open, but has now been closed. I appreciate you don't want to get into a war on a bug thread, but closing bugs without saying why confuses us who report them and want to know when they are fixed.

Concretely, to answer your questions: I want to test with GHC 8.4.1. I want to test using my CI. I want to test exactly the same way I've tested every GHC release up until now, unless someone offers a compelling reason that a better approach makes more sense. I want to test because people asked me to test. I want to know when the issue is fixed, and thus I want to subscribe to a github ticket. I want to know if you plan on fixing this in the next week or so, only once GHC 8.4 is in Beta/RC1, or at some other point - so I can plan what I do in response. I am grateful for the work everyone does maintaining the infrastructure and packages, this one included 😄

**Comment by gbaz** [comment]
> I want to test exactly the same way I've tested every GHC release up until now, unless someone offers a compelling reason that a better approach makes more sense.

This isn't every ghc release until now -- it is an early alpha release, which is a new thing, part of a new release process: https://ghc.haskell.org/trac/ghc/wiki/WorkingConventions/Releases/NewSchedule

Historically libraries were released in conjunction only with the candidate phase of a release. The new process was not intended to also have more libraries released for the alpha phase. Rather, the overlay package index is currently the recommended way to deal with it, as per: https://mail.haskell.org/pipermail/ghc-devs/2017-December/015237.html

If this wasn't the case, we'd potentially have a _ton_ more package releases, a number of which would not be particularly compatible with _any_ properly released version of GHC.

This new alpha ghc release is uncharted territory for everyone. As per the announcement this is way earlier than normal -- the fact that we can test with it at all is forward progress.

As such, this isn't a bug. It is a disagreement over the new alpha release process. Those who want to weigh in on the alpha testing process should probably do so on the ghc-devs list.

**Comment by ndmitchell** [comment]
> This isn't every ghc release until now -- it is an early alpha release

I've been tested with GHC HEAD for many years, and releasing packages to Hackage specifically to deal with GHC HEAD. To me, the process still hasn't changed (although I appreciate it has to the rest of the world).

If the decision of the primitive maintainers is to not release until GHC goes into Beta, that's fine, but the resolution of this ticket should be "Open" and the comment should be "We'll fix it when GHC goes into Beta, if you disagree email ghc-devs".

**Comment by gbaz** [comment]
Except... I think that the changes are already in HEAD of `primitive` -- there's just no release cut? i.e. https://github.com/haskell/primitive/commit/7e6e7b4667020b61986c60e0c2d642e700e7966d

**Comment by ndmitchell** [comment]
Yep, then the bug is "No released versions work with the unreleased GHC 8.4". Concretely, GHC 8.4 has compiled binaries, and I'm testing it on my CI. I have taken action to say released primitive is not working. When that stops being the case I need to undo that action. I want to "watch" something, so a ticket is mightily helpful for that. I imagine when primitive does release to Hackage then @GeorgeCo and @svenpanne are going to also take some action (e.g. additional testing, reverting workarounds etc).
